[civ2][docker/4] build ray.py38.cu118 docker image

can-anyscale opened the pull request:
As title, build ray.py38.cu118 docker image. Essentially, install ray on top of raypy38cu118base wanda image.

Test:
- CI
- Docker is generated (not uploaded yet): https://buildkite.com/ray-project/premerge/builds/4490#018a5117-099b-42f8-8d8f-13ab162723d7/129-2169

aslonnie commented on ci/ray_ci/docker_container.py:
```diff
@@ -0,0 +1,44 @@
+from ci.ray_ci.utils import docker_pull
+
+
+class DockerContainer(Container):
```
could you add some docs for this class and its methods on what this is doing?

aslonnie commented on ci/ray_ci/docker_container.py:
```diff
@@ -0,0 +1,44 @@
+        super().__init__(
+            "forge",
+            volumes=[
+                f"{os.environ['RAYCI_CHECKOUT_DIR']}:/rayci",
```
could you read out `RAYCI_CHECKOUT_DIR` and throw an exception if it is empty?

aslonnie commented on ci/ray_ci/docker_container.py:
```diff
@@ -0,0 +1,44 @@
+
+    def run(self) -> None:
+        base_image = (
+            f"{_DOCKER_ECR_REPO}:{os.environ['RAYCI_BUILD_ID']}-raypy38cu118base"
```
same here, throw an exception?

aslonnie commented on ci/ray_ci/docker_container.py:
```diff
@@ -0,0 +1,44 @@
+            f"{PYTHON_VERSIONS[self.python_version]['bin_path']}-"
+            "manylinux2014_x86_64.whl"
+        )
+        constraints_file = (
+            "requirements_compiled_py37.txt"
+            if self.python_version == "py37"
+            else "requirements_compiled.txt"
+        )
```
I always find this a bit hard to understand.. could you change it to a plain:

```
if xxx:
  x = one
else:
  x = two
```

it even take fewer lines..

or 

```
x = default_value
if xxx: x = alternative_value;
```

aslonnie commented on ci/ray_ci/docker_container.py:
```diff
@@ -0,0 +1,44 @@
+        docker_pull(base_image)
+        wheel_name = (
+            "ray-3.0.0.dev0-"
+            f"{PYTHON_VERSIONS[self.python_version]['bin_path']}-"
```
get a var for `PYTHON_VERSIONS[self.python_version]['bin_path']` and try to get the string to a single line?

also in other places.

I think in general, I find it easier if the thing embedded in the f string is closer to a single var reference.

aslonnie commented on ci/ray_ci/docker_container.py:
```diff
@@ -0,0 +1,44 @@
+            f"{self.python_version}-cu118"
+        )
+        self.run_script(
+            [
+                "./ci/build/build-ray-docker.sh "
+                f"{wheel_name} {base_image} {constraints_file} {ray_image}"
+            ]
```
this is a string list with just one string?

could you remind me why it takes a list?

can-anyscale replied:
It is a list of command that is written in one string separated by \n since it is passed in to "bash -c"

I used to use

"""
cmd_a
cmd_b
"""

but that needs to break python indentation so it looks weird, I ended up provide an API that accepts a list of commands

aslonnie commented on ci/ray_ci/test_docker_container.py:
```diff
@@ -0,0 +1,36 @@
+
+    with mock.patch.dict(
+        os.environ,
+        {
+            "RAYCI_CHECKOUT_DIR": "/ray",
+            "RAYCI_BUILD_ID": "123",
+            "RAYCI_WORK_REPO": "rayproject/citemp",
+            "BUILDKITE_COMMIT": "123456",
+        },
```
is it possible to send this dict in a config on init? and have a function to gather these from os env? so that we do not need to do lots of mocking here in testing?

we probably should talk more about how to reduce the amount of mocking that is being used.

can-anyscale replied:
totally

aslonnie commented on ci/build/build-ray-docker.sh:
```diff
@@ -0,0 +1,24 @@
+
+cd "${CPU_TMP}"
+tar --mtime="UTC 2020-01-01" -c -f - . \
+    | docker build --progress=plain \
+        --build-arg FROM="$SOURCE_IMAGE" \
+        --build-arg WHEEL_PATH=".whl/${WHEEL_NAME}" \
+        --build-arg CONSTRAINTS_FILE="$CONSTRAINTS_FILE" \
+        -t "$DEST_IMAGE" -f Dockerfile -
```
should we use wanda here? so you do not need to `tar` command, because wanda will handle that.

can-anyscale replied:
possibly, but where is a follow up logic afterwards for generating aliases and publishing docker image which happens outside of wanda so it needs a subsequent buildkite job, etc; or wanda needs to support a post-command feature, both seems complex

also the resulting docker image will always be different so less cache benefit from wanda

aslonnie replied:
you can use wanda just as a stand-alone command to build an image.

aslonnie replied:
it does not have to be used with rayci.

can-anyscale commented on docker/ray/Dockerfile:
```diff
@@ -1,5 +1,6 @@
 ARG BASE_IMAGE
-FROM rayproject/ray-deps:nightly"$BASE_IMAGE"
+ARG FROM=rayproject/ray-deps:nightly"$BASE_IMAGE"
```
Ideally I don't want to change civ1 code if possible, I think Docker might accept something like:

ARG BASE_IMAGE
ARG FULL_BASE_IMAGE
FROM ${FULL_BASE_IMAGE-rayproject/ray-deps:nightly"$BASE_IMAGE"}

can-anyscale commented:
@aslonnie: addressed all comments, except for the usage of wanda, I cannot figure that self on my own, is there an example I can use (run_wanda.sh ??), thankks

aslonnie reviewed (approved):
just some small things.
